src/System.Linq/src/System/Linq/Concatenate.cs

```diff
@@ -45,5 +32,227 @@ private static IEnumerable<TSource> PrependIterator<TSource>(IEnumerable<TSource
             yield return element;
             foreach (TSource e1 in source) yield return e1;
         }
+
+        public static IEnumerable<TSource> Concat<TSource>(this IEnumerable<TSource> first, IEnumerable<TSource> second)
+        {
+            if (first == null) throw Error.ArgumentNull("first");
+            if (second == null) throw Error.ArgumentNull("second");
+
+            var concatFirst = first as ConcatIterator<TSource>;
+            return concatFirst != null ?
+                UpgradeConcatIterator(concatFirst, second) :
+                new Concat2Iterator<TSource>(first, second);
+        }
+
+        private static ConcatIterator<TSource> UpgradeConcatIterator<TSource>(ConcatIterator<TSource> first, IEnumerable<TSource> second)
+        {
+            Debug.Assert(first is Concat2Iterator<TSource> || first is Concat3Iterator<TSource> || first is ConcatNIterator<TSource>);
+            Debug.Assert(second != null);
+
+            // Are we upgrading from two sources to three?
+            Concat2Iterator<TSource> two = first as Concat2Iterator<TSource>;
+            if (two != null)
+            {
+                return new Concat3Iterator<TSource>(two._first, two._second, second);
+            }
+
+            // From three to four?
+            Concat3Iterator<TSource> three = first as Concat3Iterator<TSource>;
+            if (three != null)
+            {
+                return new ConcatNIterator<TSource>(three._first, three._second, three._third, second);
+            }
+
+            // From four+ to one more
+            const int MaxLengthForConcatArray = 64; // arbitrary limit on array size to avoid lots of large array allocations
+            ConcatNIterator<TSource> n = (ConcatNIterator<TSource>)first;
+            if (n._sources.Length < MaxLengthForConcatArray)
+            {
+                var sources = new IEnumerable<TSource>[n._sources.Length + 1];
+                Array.Copy(n._sources, 0, sources, 0, n._sources.Length);
+                sources[n._sources.Length] = second;
+                return new ConcatNIterator<TSource>(sources);
+            }
```

Comment:
I realized I can likely avoid these array allocations by chaining the concat iterators and changing how GetEnumerable is implemented. I'll try it out tomorrow. Not sure it'll better or not.